fix return page

tuhalang opened the pull request:
This PR resolve issue #383

barisere commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,106 @@
+		}
+
+		filterSelector.txCount, filterSelector.txCountErr = wallet.TransactionCount(walletcore.BuildTransactionFilter(filter))
+		if filterSelector.txCountErr != nil {
```
This technique is returning an error implicitly. The function signature does not indicate the possibility of an error, so the caller has to be aware of the `FilterSelector.txCountErr` property. Can you make the error condition explicit?

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -77,10 +77,30 @@ func (handler *HistoryHandler) Render(window *nucular.Window) {
+			handler.selectedFilter = selectedFilter
+			handler.filter = filter
+			handler.transactions = nil
+			go handler.fetchTransactions(filter)
```
Wallet operations should not be performed directly from the `Render` function, use the `BeforeRender` function instead. `Render` should only be for drawing/rendering views on the window; it is called repeatedly whenever there's a slight action such as moving the mouse. As a matter of fact, a single mouse move could cause the `Render` function to be called multiple times in a second.

The only safe places to call wallet functions is in `BeforeRender` and in event listeners such as button click event listeners.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -9,13 +9,19 @@ import (
+	filter               *txindex.ReadFilter
+
 	totalTxCount int
+	txCountErr               error
```
go fmt

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -32,34 +38,54 @@ type HistoryHandler struct {
 	handler.isFetchingTransactions = false
 
-	handler.clearTxDetails()
+	handler.filterSelectorWidget, handler.txCountErr = widgets.FilterSelectorWidget(wallet, nil)
```
This call is rendered obsolete by similar function call in line 57. Please delete.

Sirmorrison replied:
Not exactly... Line 47 generates the initial receive address on the first user account. The call on line 57 only comes to effect only after the dropdown option is selected or changed.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -32,34 +38,54 @@ type HistoryHandler struct {
 	}
 
+	// fetch filtered data
+	handler.filterSelectorWidget, handler.txCountErr = widgets.FilterSelectorWidget(wallet, func(int){
```
Why is the error returned from the `widgets.FilterSelectorWidget` function assigned to `txCountErr`? The relattionship between tx count and FilterSelectorWidget (if such a relationship even exists) is not immediately obvious.

Should probably rename `txCountErr` to `filterSelectorErr`.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -32,34 +38,54 @@ type HistoryHandler struct {
+
+		if selectedFilter != handler.selectedFilter {
+			handler.selectedFilter = selectedFilter
+			handler.filter = filter
```
Rename `filter` and `handler.filter` to `selectedTxFilter` and `handler.selectedTxFilter` respectively. Also rename `selectedFilter` and `handler.selectedFilter` to `currentFilterText` and `handler.currentFilterText` respectively.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -32,34 +38,54 @@ type HistoryHandler struct {
+	// fetch filtered data
+	handler.filterSelectorWidget, handler.txCountErr = widgets.FilterSelectorWidget(wallet, func(int){
+		totalTxCount, filter, selectedFilter, err := handler.filterSelectorWidget.GetSelectedFilter()
+		handler.totalTxCount = totalTxCount
```
Does the `totalTxCount` count change? `handler.totalTxCount` is already assigned in line 50; don't see why it needs to be returned by `filterSelectorWidget.GetSelectedFilter()`.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -32,34 +38,54 @@ type HistoryHandler struct {
+		totalTxCount, filter, selectedFilter, err := handler.filterSelectorWidget.GetSelectedFilter()
+		handler.totalTxCount = totalTxCount
+		if err != nil {
+			handler.fetchHistoryError = err
```
`err` isn't an history fetch error, is it?

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -32,34 +38,54 @@ type HistoryHandler struct {
+		}
+	})
+
+	handler.clearTxDetails()
```
You want to clear tx details before calling `fetchTransactions`. Restore this `handler.clearTxDetails()` call to after line 45.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -32,34 +38,54 @@ type HistoryHandler struct {
 	handler.refreshWindowDisplay() // refresh display to show loading indicator
 
 	txHistoryOffset := 0
+
```
This space is unnecessary. The next set of lines directly relate to the preceding line.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -77,10 +103,18 @@ func (handler *HistoryHandler) Render(window *nucular.Window) {
+		handler.filterSelectorWidget.Render(contentWindow)
+
+		if handler.txCountErr != nil {
+			contentWindow.DisplayErrorMessage("Error loading history", handler.fetchHistoryError)
```
How is a tx **COUNT** error same as an error loading history? Oh by the way, you're referencing `handler.fetchHistoryError` in the if statement body but your if statement checked for `handler.txCountErr != nil`

itswisdomagain commented on nuklear/widgets/accountselector.go:
```diff
@@ -66,7 +66,7 @@ func (accountSelector *AccountSelector) Render(window *Window, addColumns ...int
 	if accountSelector.accountsFetchError != nil {
 		window.DisplayErrorMessage("Fetch accounts error", accountSelector.accountsFetchError)
-	} else if len(accountSelector.accounts) == 1 {
+	} else if len(accountSelector.accounts) != 1 {
```
Why is this change necessary?

itswisdomagain commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,135 @@
+	wallet walletcore.Wallet
+
+	filters                  []string
+	filter                   string
```
`selectedFilter`

itswisdomagain commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,135 @@
+	changed                  bool
+	selectedFilterAndCount   string
+	txFilter                 *txindex.ReadFilter
+	selectionChanged         func(int)
```
The `int` is not needed obviously, can just use a simple `func()`.

itswisdomagain commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,135 @@
+
+	filterSelector.filters = walletcore.TransactionFilters
+
+	filterSelector.transactionCountByFilter = make([]string, len(filterSelector.filters))
```
The empty line spaces at lines 41 and 43 can be deleted without hurting readability.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -9,13 +9,19 @@ import (
 
+	filterSelectorWidget *widgets.FilterSelector
+	selectedFilter       string
+	filter               *txindex.ReadFilter
```
Shouldn't need to save this variable in this struct. Only save the selected filter text and whenever you need the readfilter, you call `walletcore.BuildTransactionFilter()`

Sirmorrison replied:
`filter` this is just a variable stored here oh not the filters themselves.

itswisdomagain replied:
I did not debate that with you. Please read my comment again and apply accordingly.

itswisdomagain commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,135 @@
+			filterSelector.totalTxCount = filterSelector.txCount
+		}
+
+		filterSelector.txCount, filterSelector.txCountErr = wallet.TransactionCount(walletcore.BuildTransactionFilter(filter))
```
Each time this loop runs, `filterSelector.txCount` is re-assigned. The final value of `filterSelector.txCount` is the value set by the last loop run. I believe this variable does not need to be a property of the `filterSelector` struct, but can be a local variable used here and here alone.

itswisdomagain commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,135 @@
+	filters                  []string
+	filter                   string
+	txCount                  int
+	transactionCountByFilter []string
```
Why is a transaction COUNT slice a string slice instead of a number slice? Probably need to rename your variable.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -9,30 +9,35 @@ import (
+	selectedTxFilter      	*txindex.ReadFilter
+
+	totalTxCount 			int
+	filterSelectorErr       error
```
- go fmt please
- move this `filterSelectorErr` property close to the `filterSelectorWidget` property. Doesn't that just make more sense?

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -9,30 +9,35 @@ import (
+
+	filterSelectorWidget 	*widgets.FilterSelector
+	currentFilterText    	string
+	selectedTxFilter      	*txindex.ReadFilter
```
I've mentioned in a previous review that this does not need to be stored. Only `currentFilterText` needs to be stored.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -41,25 +46,44 @@ func (handler *HistoryHandler) BeforeRender(wallet walletcore.Wallet, settings *
 	}
 
+	// fetch filtered data
+	handler.filterSelectorWidget, handler.filterSelectorErr = widgets.FilterSelectorWidget(wallet, func(){
```
This code in line 57 does not fetch filtered data. It sets up the filter widget. Update your comment.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -41,25 +46,44 @@ func (handler *HistoryHandler) BeforeRender(wallet walletcore.Wallet, settings *
 
+	// fetch filtered data
+	handler.filterSelectorWidget, handler.filterSelectorErr = widgets.FilterSelectorWidget(wallet, func(){
+		_, selectedTxFilter, currentFilterText, err := handler.filterSelectorWidget.GetSelectedFilter()
```
`handler.filterSelectorWidget.GetSelectedFilter` should only return `currentFilterText, err` or just `currentFilterText` even.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -77,10 +101,18 @@ func (handler *HistoryHandler) Render(window *nucular.Window) {
 	widgets.PageContentWindowDefaultPadding("History", window, func(contentWindow *widgets.Window) {
-		// show transactions first, if any
-		if len(handler.transactions) > 0 {
+		handler.filterSelectorWidget.Render(contentWindow)
```
`handler.filterSelectorWidget` may be nil if line 57 returns an error. You should check that it is not nil before rendering it here.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -41,25 +46,44 @@ func (handler *HistoryHandler) BeforeRender(wallet walletcore.Wallet, settings *
+	handler.filterSelectorWidget, handler.filterSelectorErr = widgets.FilterSelectorWidget(wallet, func(){
+		_, selectedTxFilter, currentFilterText, err := handler.filterSelectorWidget.GetSelectedFilter()
+		if err != nil {
+			handler.filterSelectorErr = err
```
This is not an error with the filter selector. If this code here executes, then it means the filter has been successfully rendered on the page. This sub-function only gets triggered if a user selects an item from the filter dropdown. This is probably more related to `fetchHistoryError` since the next thing after a user selects an item from the filter is an attempt to load transactions for the selected filter item.

Sirmorrison replied:
https://github.com/raedahgroup/godcr/pull/384#discussion_r327695663

already resolves this as no error is returned.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -9,30 +9,33 @@ import (
+	filterSelectorErr       error
+	currentFilterText    	string
+
+	totalTxCount 			int
```
go fmt please

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -41,25 +44,42 @@ func (handler *HistoryHandler) BeforeRender(wallet walletcore.Wallet, settings *
+	handler.filterSelectorWidget, handler.filterSelectorErr = widgets.FilterSelectorWidget(wallet, func(){
+		currentFilterText, err := handler.filterSelectorWidget.GetSelectedFilter()
+		if err != nil {
+			handler.fetchHistoryError = err
```
Again you mark a comment as resolved without actually resolving it or commenting on it. Please go back to https://github.com/raedahgroup/godcr/pull/384/files#r327002260

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -41,25 +44,42 @@ func (handler *HistoryHandler) BeforeRender(wallet walletcore.Wallet, settings *
+			handler.fetchHistoryError = err
+			return
+		}
+		if currentFilterText != handler.currentFilterText {
```
Empty line space before this.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -41,25 +44,42 @@ func (handler *HistoryHandler) BeforeRender(wallet walletcore.Wallet, settings *
+			go handler.fetchTransactions(walletcore.BuildTransactionFilter(handler.currentFilterText))
+		}
+
+		refreshWindowDisplay()
```
This should only be called after the transactions are fetched, i.e. from inside the `handler.fetchTransactions` method. Please remove from here.

If you need to refresh the display, that should be after line 63 where you clear the previous list of transactions.

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -77,10 +97,20 @@ func (handler *HistoryHandler) Render(window *nucular.Window) {
-		// show transactions first, if any
-		if len(handler.transactions) > 0 {
+		if handler.fetchHistoryError == nil {
+			handler.filterSelectorWidget.Render(contentWindow)
```
How does `handler.fetchHistoryError` relate to `handler.filterSelectorWidget`?

itswisdomagain commented on nuklear/pagehandlers/history.go:
```diff
@@ -77,10 +97,20 @@ func (handler *HistoryHandler) Render(window *nucular.Window) {
+		}
+
+		if len(handler.transactions) == 0 {
+			contentWindow.AddWrappedLabel("No Transactions to display yet", widgets.CenterAlign)
```
Transactions -> transactions

itswisdomagain commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,128 @@
+		}
+
+		txCount, txCountErr := wallet.TransactionCount(walletcore.BuildTransactionFilter(filter))
+		filterSelector.txCountErr = txCountErr
```
This assignment is unncessary. Just return the error!

itswisdomagain commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,128 @@
+
+	for index, filter := range filterSelector.filters {
+		if filter == "All" {
+			txCount, txCountErr := wallet.TransactionCount(nil)
```
If filter is all, you call `wallet.TransactionCount` twice. This is unnecessary. See the similar comment on your other PR #383.

itswisdomagain commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,128 @@
+			return nil, filterSelector.txCountErr
+		}
+
+		filterSelector.transactionFilter[index] = fmt.Sprintf("%s (%d)", filter, txCount)
```
You should ignore 0 count filters. This is done better in #383

itswisdomagain commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,128 @@
+	// row with fixed column widths to hold account selection prompt, the account widget, and any other widgets that may be added later
+	rowColumns := make([]int, 2)
+	rowColumns[0] = window.LabelWidth("filter")
+	rowColumns[1] = filterSelectorWidth
```
lines 64 to 66 can be simplified to 
```
rowColumns := []int {
     window.LabelWidth("filter"),
     filterSelectorWidth,
}
```

By the way, those are not columns! Please use proper variable names!

itswisdomagain commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,128 @@
+type FilterSelector struct {
+	wallet walletcore.Wallet
+	filters                  []string
+	transactionFilter []string
```
- `transactionFilters`, this is a collection of items, not a single item
- what's the difference between `filters` and `transactionFilters`?
- go fmt

Sirmorrison commented on nuklear/widgets/dropDown.go:
```diff
@@ -0,0 +1,128 @@
+	filterSelectorWidth := defaultFilterSelectorWidth
+
+	// row with fixed column widths to hold account selection prompt, the account widget, and any other widgets that may be added later
+	rowColumns := []int {
```
here

oshorefueled commented on ui/overview_page.go:
```diff
@@ -679,6 +679,8 @@ func (pg *overviewPage) Handler(eq event.Queue, c pageCommon) {
 				txn := (*pg.walletTransactions).Recent[index]
 				*pg.walletTransaction = &txn
+				*c.returnPage = PageOverview
+				c.setReturnPage(PageOverview)
```
This looks like the returnPage field is being set twice. It is set directly on line 682 once and a second time with a method 683.

oshorefueled commented on ui/transactions_page.go:
```diff
@@ -260,11 +260,14 @@ func (pg *transactionsPage) sortTransactions(common *pageCommon) {
-			changePage(PageTransactionDetails)
+
+			*common.returnPage = PageTransactions
+			common.setReturnPage(PageTransactions)
```
This looks like the returnPage field is being set twice. It is set directly on line 268 once and a second time with a method 269.

tuhalang committed:
Fix back button on transactions detail page

- The detail of a transaction can be seen from overview page. However, the back button on transactions detail always navigates to the transactions page
- I changed it to be if viewed from the overview page it will return to the overview page
